### ISSUE TITLE
Sample for Shifting data source

### ISSUE THREAD

**Issue by valterc** [hammerheadnav/karoo-ext#14]
Hi, I'm trying to create a connection to a trainer with custom gear shifting and I want to add a new extension that reports Shifting Gear information to Karoo.

I have the following code (simplified for example):

```kt
class KarooExtService : KarooExtension(BuildConfig.APPLICATION_ID, BuildConfig.VERSION_NAME) {

override fun startScan(emitter: Emitter<Device>) {
        emitter.onNext(
            Device(
                extension,
                "1",
                listOf(
                    DataType.Source.SHIFTING_FRONT_GEAR,
                    DataType.Source.SHIFTING_REAR_GEAR
                ),
                "1"
            )
        )
    }

override fun connectDevice(uid: String, emitter: Emitter<DeviceEvent>) {
        val job = CoroutineScope(Dispatchers.IO).launch {
            emitter.onNext(OnConnectionStatus(ConnectionStatus.SEARCHING))
            delay(2000)

            emitter.onNext(OnConnectionStatus(ConnectionStatus.CONNECTED))
            delay(1000)

            emitter.onNext(OnBatteryStatus(BatteryStatus.NEW))
            delay(1000)

            emitter.onNext(OnManufacturerInfo(ManufacturerInfo("Test", "Test", "Test")))
            delay(1000)

            repeat(Int.MAX_VALUE) {
                emitter.onNext(
                    OnDataPoint(
                        DataPoint(
                            DataType.Type.SHIFTING_FRONT_GEAR,
                            values = mapOf(
                                DataType.Field.SHIFTING_FRONT_GEAR to 2.0,
                                DataType.Field.SHIFTING_FRONT_GEAR_MAX to 2.0),
                            sourceId = "1",
                        ),
                    )
                )
                emitter.onNext(
                    OnDataPoint(
                        DataPoint(
                            DataType.Type.SHIFTING_REAR_GEAR,
                            values = mapOf(
                                DataType.Field.SHIFTING_REAR_GEAR to 2.0,
                                DataType.Field.SHIFTING_REAR_GEAR_MAX to 2.0),
                            sourceId = "1",
                        ),
                    )
                )
                delay(1000)
            }
            awaitCancellation()
        }

        emitter.setCancellable {
            job.cancel()
        }
    }

}
```

With the code above I can add the extension sensor, it is recognized as a 'Shifting' sensor, it connects, but the data types related show a `0` and I get invalid data point logs:

```
HHApp                   io.hammerhead.sensorservice          I  SensorPairing: Connection state for Device(uid=EXT::10100-129-5, info=DeviceInfo(name=10100-129-5, axsBikeId=null, syncedWithAxs=false, nameEditedByUser=false, lastBatteryValue=BatteryValue(batteryStatus=NEW, percentage=-1, fetchTime=1733151296917, components=null), manufacturerInfo=ManufacturerInfo(manufacturer=Test, serialNumber=Test, modelNumber=Test, hardwareVersion=null, softwareVersion=null, components=null), componentInfo=null), supportedDataTypes=[s2.b@28e6d8d, s2.g@413cb42, s2.j@b67d753, s2.f@c042890, s2.i@1815589], blockedDataTypes=[], isEnabled=true, isConnectedAtLeastOnce=true, connectionType=EXTENSION, wheelCircumferenceInfo=null, shiftingInfo=null, lightMode=null, autoLightControl=null, ctfOffset=null, priority=4, functionSets=null, axsPrimarySerial=null, alerts=[CONFIGURE_GEARS]) changed to DISCONNECTED
HHApp                   io.hammerhead.sensorservice          I  SensorPairing: Connecting to Device(uid=EXT::10100-129-5, info=DeviceInfo(name=10100-129-5, axsBikeId=null, syncedWithAxs=false, nameEditedByUser=false, lastBatteryValue=BatteryValue(batteryStatus=NEW, percentage=-1, fetchTime=1733151296917, components=null), manufacturerInfo=ManufacturerInfo(manufacturer=Test, serialNumber=Test, modelNumber=Test, hardwareVersion=null, softwareVersion=null, components=null), componentInfo=null), supportedDataTypes=[s2.b@28e6d8d, s2.g@413cb42, s2.j@b67d753, s2.f@c042890, s2.i@1815589], blockedDataTypes=[], isEnabled=true, isConnectedAtLeastOnce=true, connectionType=EXTENSION, wheelCircumferenceInfo=null, shiftingInfo=null, lightMode=null, autoLightControl=null, ctfOffset=null, priority=4, functionSets=null, axsPrimarySerial=null, alerts=[CONFIGURE_GEARS])
HHApp                   io.hammerhead.sensorservice          I  SensorPairing: Connection state for Device(uid=EXT::10100-129-5, info=DeviceInfo(name=10100-129-5, axsBikeId=null, syncedWithAxs=false, nameEditedByUser=false, lastBatteryValue=BatteryValue(batteryStatus=NEW, percentage=-1, fetchTime=1733151296917, components=null), manufacturerInfo=ManufacturerInfo(manufacturer=Test, serialNumber=Test, modelNumber=Test, hardwareVersion=null, softwareVersion=null, components=null), componentInfo=null), supportedDataTypes=[s2.b@28e6d8d, s2.g@413cb42, s2.j@b67d753, s2.f@c042890, s2.i@1815589], blockedDataTypes=[], isEnabled=true, isConnectedAtLeastOnce=true, connectionType=EXTENSION, wheelCircumferenceInfo=null, shiftingInfo=null, lightMode=null, autoLightControl=null, ctfOffset=null, priority=4, functionSets=null, axsPrimarySerial=null, alerts=[CONFIGURE_GEARS]) changed to SEARCHING
HHApp                   io.hammerhead.activityservice        I  StreamFetching: Unsubscribing from TYPE_SHIFTING_FRONT_GEAR_ID-EXT::10100-129-5 streamObservable
HHApp                   io.hammerhead.activityservice        I  StreamFetching: Unsubscribing from TYPE_SHIFTING_REAR_GEAR_ID-EXT::10100-129-5 streamObservable
HHApp                   io.hammerhead.sensorservice          I  Extensions: Device EXT::10100-129-5 connected
HHApp                   io.hammerhead.sensorservice          I  SensorPairing: Connection state for Device(uid=EXT::10100-129-5, info=DeviceInfo(name=10100-129-5, axsBikeId=null, syncedWithAxs=false, nameEditedByUser=false, lastBatteryValue=BatteryValue(batteryStatus=NEW, percentage=-1, fetchTime=1733151296917, components=null), manufacturerInfo=ManufacturerInfo(manufacturer=Test, serialNumber=Test, modelNumber=Test, hardwareVersion=null, softwareVersion=null, components=null), componentInfo=null), supportedDataTypes=[s2.b@28e6d8d, s2.g@413cb42, s2.j@b67d753, s2.f@c042890, s2.i@1815589], blockedDataTypes=[], isEnabled=true, isConnectedAtLeastOnce=true, connectionType=EXTENSION, wheelCircumferenceInfo=null, shiftingInfo=null, lightMode=null, autoLightControl=null, ctfOffset=null, priority=4, functionSets=null, axsPrimarySerial=null, alerts=[CONFIGURE_GEARS]) changed to CONNECTED
HHApp                   io.hammerhead.sensorservice          W  Device(uid=EXT::10100-129-5, info=DeviceInfo(name=10100-129-5, axsBikeId=null, syncedWithAxs=false, nameEditedByUser=false, lastBatteryValue=BatteryValue(batteryStatus=NEW, percentage=-1, fetchTime=1733151296917, components=null), manufacturerInfo=ManufacturerInfo(manufacturer=Test, serialNumber=Test, modelNumber=Test, hardwareVersion=null, softwareVersion=null, components=null), componentInfo=null), supportedDataTypes=[s2.b@28e6d8d, s2.g@413cb42, s2.j@b67d753, s2.f@c042890, s2.i@1815589], blockedDataTypes=[], isEnabled=true, isConnectedAtLeastOnce=true, connectionType=EXTENSION, wheelCircumferenceInfo=null, shiftingInfo=null, lightMode=null, autoLightControl=null, ctfOffset=null, priority=4, functionSets=null, axsPrimarySerial=null, alerts=[CONFIGURE_GEARS]) supported data types observable giving up
HHApp                   io.hammerhead.activityservice        I  StreamFetching: Subscribing to TYPE_SHIFTING_REAR_GEAR_ID-EXT::10100-129-5 streamObservable
HHApp                   io.hammerhead.activityservice        I  StreamFetching: Subscribing to TYPE_SHIFTING_FRONT_GEAR_ID-EXT::10100-129-5 streamObservable
HHApp                   io.hammerhead.activityservice        I  Ride: Filtering invalid datapoint DataPoint(values={FIELD_SHIFTING_FRONT_GEAR_ID=Value(value=2.0, format=2), FIELD_SHIFTING_REAR_GEAR_ID=Value(value=2.0, format=2)}, sampleDuration=1000, timeStamp=1733152060356, dataType=hhm3.i@2ea3e91, deviceSourceIds=[EXT::10100-129-5], metaFields={}, persistentState={}); hasMandatoryFields=true noExtraFields=true validValues=false
HHApp                   io.hammerhead.activityservice        I  Ride: Filtering invalid datapoint DataPoint(values={FIELD_SHIFTING_REAR_GEAR_ID=Value(value=2.0, format=2)}, sampleDuration=1000, timeStamp=1733152060364, dataType=hhm3.l@1f301f6, deviceSourceIds=[EXT::10100-129-5], metaFields={}, persistentState={}); hasMandatoryFields=true noExtraFields=true validValues=false
HHApp                   io.hammerhead.activityservice        I  Ride: Filtering invalid datapoint DataPoint(values={FIELD_SHIFTING_FRONT_GEAR_ID=Value(value=2.0, format=2), FIELD_SHIFTING_REAR_GEAR_ID=Value(value=2.0, format=2)}, sampleDuration=1000, timeStamp=1733152060364, dataType=hhm3.i@2ea3e91, deviceSourceIds=[EXT::10100-129-5], metaFields={}, persistentState={}); hasMandatoryFields=true noExtraFields=true validValues=false
HHApp                   io.hammerhead.activityservice        I  Ride: Filtering invalid datapoint DataPoint(values={FIELD_SHIFTING_FRONT_GEAR_ID=Value(value=2.0, format=2), FIELD_SHIFTING_FRONT_GEAR_MAX_ID=Value(value=2.0, format=2)}, sampleDuration=1000, timeStamp=1733157581786, dataType=hhm3.h@670c14, deviceSourceIds=[EXT::10100-129-5], metaFields={}, persistentState={}); hasMandatoryFields=true noExtraFields=true validValues=false
```

Would it be possible to have a sample for such data source extension sensor?

Thank you.

**Comment by brentnd** [comment]
This is an defect in the data structures for how values are provided by karoo-extensions library in `OnDataPoint`.

As the error notes, in `validValues=false` this is checking types of the fields provided. Unlike the sample (HR) and other validated extension types (power, speed, cadence), the fields for shifting data types use integer values internally and the doubles provided by the extension are failing to validate.

This will need to be fixed on the Karoo-side by mapping to the appropriate internal type instead of always using a double while the karoo-ext contract can remain as a double.

**Comment by valterc** [comment]
Thanks for the quick follow up.

If it would be possible, could the team also add support (internal mapping) for emitting the `SHIFTING_FRONT_GEAR_TEETH` and `SHIFTING_REAR_GEAR_TEETH` together in the respective data points (`SHIFTING_FRONT_GEAR`, `SHIFTING_REAR_GEAR`)?

> This will need to be fixed on the Karoo-side by mapping to the appropriate internal type instead of always using a double while the karoo-ext contract can remain as a double.

On that note, is there any plan to support other data types, for example, in the existing model, there are other fields that don't map well to double/integer:
- [DataType.Field.SHIFTING_MODE](https://github.com/hammerheadnav/karoo-ext/blob/efac75b5f1b1fdf9f792eeffbf2dec92e48cdd65/lib/src/main/kotlin/io/hammerhead/karooext/models/DataType.kt#L1082)
- [DataType.Source.SHIFTING_BATTERY](https://github.com/hammerheadnav/karoo-ext/blob/efac75b5f1b1fdf9f792eeffbf2dec92e48cdd65/lib/src/main/kotlin/io/hammerhead/karooext/models/DataType.kt#L66C19-L66C35) / [DataType.Field.SHIFTING_BATTERY_STATUS](https://github.com/hammerheadnav/karoo-ext/blob/efac75b5f1b1fdf9f792eeffbf2dec92e48cdd65/lib/src/main/kotlin/io/hammerhead/karooext/models/DataType.kt#L1079C19-L1079C42)
- (and some others not related with shifting)

Those fields seem to require different value types (`string` I assume), or is there some sort of mapping, for example battery status (0 = CRIT, 1 = LOW, 2 = OK, ...)?

**Comment by brentnd** [comment]
`FIELD_SHIFTING_FRONT_*_TEETH` fields are optional but can be provided in the values for their respective data points.

`SHIFTING_MODE` is a deprecated type that is no longer exposed in the UI so can't be supported. I'll remove it from the `Field` object.

`SHIFTING_BATTERY` as a source is supported but the fields lack documentation. I'll update to include with the source but to also list here:
* `FIELD_SHIFTING_BATTERY_STATUS_ID` is an optional integer with values from `BatteryStatus` ordinal: NEW, GOOD, OK, LOW, CRITICAL, INVALID. This can be included in front and rear data points where it applies.

Ideally we can avoid string values altogether but maybe a need will arise at some point.

**Comment by brentnd** [comment]
I also want to look into leveraging the existing "Configure Gearing" screen in sensor settings to use those teeth if they aren't provided by the extension.

**Comment by brentnd** [comment]
This is fixed in release `1.527.2014`. The fix was internal to the Karoo system code so no changes in this repo are needed once a device has that software. #16 is related to later comments here but not required for the integer value fix.
